### PR TITLE
Address issues in algorithms identified by static analysis

### PR DESCRIPTION
**Pull request by timmiesmith**
Static analysis has identified a function marked as noexcept where the SYCL code called may through an exception.  A separate issue identified is the use of an object after it was forwarded.  This PR addresses both issues.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/numeric_ranges_impl_hetero.h [hunk 1]
```diff
@@ -93,6 +93,7 @@ __pattern_transform_scan_base(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Rang
 {
     if (__rng1.empty())
         return 0;
+    oneapi::dpl::__internal::__difference_t<_Range2> __rng1_size = __rng1.size();
```
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/numeric_ranges_impl_hetero.h:96]: Should this type be `oneapi::dpl::__internal::__difference_t<_Range1>`?

**Review by mmichel11 — APPROVED** [review]
LGTM

File: include/oneapi/dpl/pstl/hetero/numeric_ranges_impl_hetero.h (hunk 1 above)
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/numeric_ranges_impl_hetero.h:96]: I'm fine with it as is or with `auto`, which would convert during the return (but would make the code less verbose).

**Review by danhoeflinger — APPROVED** [review]
LGTM

**Pull request merged**